File: src/main/java/ru/citeck/ecos/completion/java/providers/QNamesCompletionProvider.java
package ru.citeck.ecos.completion.java.providers;

import com.intellij.codeInsight.completion.CompletionParameters;
import com.intellij.codeInsight.completion.CompletionResultSet;
import com.intellij.codeInsight.completion.JavaCompletionUtil;
import com.intellij.openapi.project.Project;
import com.intellij.patterns.ElementPattern;
import com.intellij.patterns.PlatformPatterns;
import com.intellij.psi.PsiClassType;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiType;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.util.ProcessingContext;
import org.jetbrains.annotations.NotNull;
import ru.citeck.ecos.ServiceRegistry;
import ru.citeck.ecos.completion.java.JavaCompletionProvider;
import ru.citeck.ecos.index.indexers.qname.QName;

import java.util.Set;

public class QNamesCompletionProvider implements JavaCompletionProvider {

    @Override
    public void addCompletions(
            @NotNull CompletionParameters parameters,
            @NotNull ProcessingContext context,
            @NotNull CompletionResultSet result
    ) {
        Project project = parameters.getPosition().getProject();
        PsiClassType qNamePattern = PsiType.getTypeByName(QName.QNAME_PATTERN, project, GlobalSearchScope.allScope(project));
        Set<PsiType> expectedTypes = JavaCompletionUtil.getExpectedTypes(parameters);
        if (expectedTypes == null || expectedTypes.stream().noneMatch(qNamePattern::isAssignableFrom)) {
            return;
        }
        result.addAllElements(
                ServiceRegistry.getQNameService(project).createJavaLookupElements()
        );
    }

    @Override
    public ElementPattern<? extends PsiElement> getElementPattern() {
        return PlatformPatterns.psiElement();
    }

}

File: src/main/kotlin/ru/citeck/completetion/java/QNamesCompletionProvider.kt
package ru.citeck.completetion.java

import com.intellij.codeInsight.completion.*
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.command.CommandProcessor
import com.intellij.psi.PsiType
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.util.ProcessingContext
import com.intellij.util.indexing.FileBasedIndex
import icons.EcosIcons
import ru.citeck.indexes.models.AlfNamespaceIndex
import ru.citeck.indexes.qnames.QNamesService
import ru.citeck.alfresco.QName

class QNamesCompletionProvider : CompletionProvider<CompletionParameters?>() {

    override fun addCompletions(
        parameters: CompletionParameters,
        context: ProcessingContext,
        resultSet: CompletionResultSet
    ) {

        val project = parameters.editor.project ?: return
        val qNamePattern = PsiType.getTypeByName(QName.QNAME_PATTERN, project, GlobalSearchScope.allScope(project))

        if (JavaCompletionUtil.getExpectedTypes(parameters)?.any { qNamePattern.isAssignableFrom(it) } != true) {
            return
        }

        val qnames = project.getService(QNamesService::class.java).findAll()

        val fileBasedIndex = FileBasedIndex.getInstance()
        val searchScope = GlobalSearchScope.allScope(project)
        val namespaces = mutableMapOf<String, String>()

        qnames.forEach {

            var prefix = namespaces[it.uri]
            if (prefix == null) {
                prefix = fileBasedIndex.getValues(AlfNamespaceIndex.NAME, it.uri, searchScope).firstOrNull()?.prefix
                if (prefix == null) return@forEach
                namespaces[it.uri] = prefix
            }

            resultSet.addElement(
                LookupElementBuilder
                    .create("${prefix}_${it.localName}")
                    .withIcon(EcosIcons.CiteckLogo)
                    .withTypeText("${it.jClass}.${it.jField}")
                    .withInsertHandler { context, item ->
                        ApplicationManager.getApplication().runWriteAction {
                            CommandProcessor.getInstance().runUndoTransparentAction {
                                context.document.replaceString(
                                    context.startOffset,
                                    context.tailOffset,
                                    "${it.jClass}.${it.jField}"
                                )
                            }
                        }
                    }
            )
        }

    }

}